Google Analytics Re-Implementation

JamesDLorence opened the pull request:
@jpetto - Here is the pull request for the Google Analytics re-implementation.


updates optimizely tag titles

updating html codeblock syntax

updating html codeblock syntax

Google Analytics Re-Implementation

creates template tag file

1st run of re-implementation changes

2nd round of re-implementation, end at Manifesto Quote Share

Half way through cleanup

Complete with re-implementation and cleanup

Delete deprecated Google Tag Manager docs folder

Resolved merge conflict of ga-tracking.js

jpetto commented on media/js/firefox/australis/common.js:
```diff
@@ -51,6 +51,12 @@
 
+                // Track sync button clicks
+
+                dataLayer.push({
```
For consistency, let's change these to `window.dataLayer.push`.

JamesDLorence replied:
@jpetto - Are you wanting to have `window.` appended to all the `dataLayer.push` calls?

jpetto replied:
Yeah, all `dataLayer` calls should be prefixed with `window.` for consistency.

jpetto commented on media/js/firefox/family-nav.js:
```diff
@@ -121,6 +122,13 @@ if (typeof window.Mozilla === 'undefined') {
         };
 
+        $tertiaryNavs.on('click', 'a', function() {
+            dataLayer.push({
```
`window.dataLayer` for consistency.

jpetto commented on media/js/firefox/firstrun/firstrun.js:
```diff
@@ -4,19 +4,28 @@
+                'event': 'video-pause',
+                'videoTitle': 'First Run Video'
+            });
+        }
```
I think we could shorten this code up a bit. Something like:

```
window.dataLayer.push({
    'event': 'video-' + $video[0].currentTime === $video[0].duration ? 'complete' : 'pause',
    'videoTitle': 'First Run Video'
});
```

jpetto commented on media/js/firefox/sync.js:
```diff
@@ -90,15 +84,6 @@
 
-    // Send page state to GA if it hasn't already been sent in the
-    // getConfiguration callback
-    if (syncCapable === false) {
```
It looks like this was the only use of the `syncCapable` variable, so let's remove it elsewhere in the script.

jpetto commented on media/js/mozorg/contribute/contribute-2015-ga.js:
```diff
@@ -154,9 +137,18 @@ $(function () {
+            if($(this).css('display') === 'block') {
+                selectedArea = $(this).find('select option:selected').text();
+            }
+        });
```
I think it would be better to keep the `$('.area select').on('change')` handler to store the selected `area`, then reference that value when the form is submitted. Does that make sense? Something like:

```
$('.area select').on('change', function() {
    $('#inquiry-form').data('contribute-area', this.value);
});
```

And then:

```
window.dataLayer.push({
    event: 'contribute-signup-submit',
    contributeArea: $('#inquiry-form').data('contribute-area'),
    countrySelected: $('#id_country option:selected').text()
});
```

That way we wouldn't have to look at each `.area` or rely on CSS values.

jpetto commented on media/js/mozorg/home/ga-tracking.js:
```diff
@@ -5,6 +5,15 @@
+        'data-element-location': 'nav click',
+        'data-link-type': $mainNav.data('name')
+    });
+
```
Hm, working with an array of elements like this is a bit confusing to read. Let's change it to be a bit more explicit in intent. Something like:

```
$('#nav-main-menu li a').each(function() {
    var $this = $(this);
    $this.attr({
        'data-element-location': 'nav click',
        'data-link-type': $this.data('name')
    });
});
```

alexgibson replied:
This was recently refactored out in #3287 as it was sending incorrect data to GTM. The attributes are handled now directly in the template, so doesn't need JS iterating over things again.

JamesDLorence replied:
Removing then since it was refactored out.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -57,16 +52,30 @@ $(function() {
+                            quoteNumber: Number(quoteText.slice(0, 2))
+                        });
+
+                        nav_modal($(this).hasClass('prev') ? -1 : 1);
```
Since `$(this)` is used twice, lets store that reference for better performance, like `var $this = $(this);`

jpetto commented on media/js/newsletter/form.js:
```diff
@@ -82,12 +82,11 @@ $(function () {
             $self.off('submit');
             newsletter = getNewsletterName($self);
-
+            //KEEP FOR IMPLEMENTATION
```
Can probably get rid of this comment?

jpetto commented on media/js/privacy/privacy-day.js:
```diff
@@ -31,52 +31,4 @@ $(function() {
-    // Setup GA tracking for next tab buttons
-    $('.tip-footer .next a').on('click', function(e) {
-        e.preventDefault();
-        pager.nextPageWithAnimation();
```
We need to keep these event handlers for this line of code to allow the user to cycle through the different tabs/pages.

jpetto commented:
A few spots to tidy up, but definitely more manageable than the last go around. :grin:

After you've addressed the comments, amend the commit and clean up the message to be simpler and more direct. Maybe like "Refactor/simplify GA/GTM tracking."

If you get stuck reconciling the `legal-docs` submodule diff, let me know and I'll try to help.

alexgibson commented on media/js/mozorg/home/ga-tracking.js.REMOTE.82408.js:
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This seems to be a duplicate file - was it intended to be included?

alexgibson commented on media/js/mozorg/home/ga-tracking.js.LOCAL.82408.js:
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This seems to be a duplicate file - was it intended to be included?

alexgibson commented on media/js/mozorg/home/ga-tracking.js.BASE.82408.js:
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This seems to be a duplicate file - was it intended to be included?

alexgibson commented on media/js/mozorg/home/ga-tracking.js.BACKUP.82408.js:
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This seems to be a duplicate file - was it intended to be included?

JamesDLorence commented:
@jpetto  - All these changes are very straight forward. I am not sure what the issue with the `legal-docs` is though and how to reconcile or what needs to be reconciled. Could you help? I am able to hop on a quick call if that would be easier.

jpetto commented:
Hey @JamesDLorence - Can you rebase against `bedrock/master` to resolve the merge conflicts? Once that's complete, I'll begin review.

Thanks!

jpetto commented on media/js/base/global.js:
```diff
@@ -60,12 +60,15 @@ function init_android_download_links() {
-            event: 'change-language',
-            languageSelected: $(this).val()
+            'event': 'change-language',
+            'languageSelected': $(this).val(),
```
Since we already have it cached, use `$language.val()` here instead.

jpetto commented on media/js/mozorg/contribute/contribute-2015-ga.js:
```diff
@@ -121,27 +125,33 @@ $(function () {
+    // Track contribute area on form submission
+    $('.area select').on('change', function() {
+        $('#inquiry-form').data('contribute-area', this.value);
+    });
```
Isn't this same thing happening on lines 140-142 above?

JamesDLorence replied:
Yep. Removed 150-153.

jpetto commented on media/js/mozorg/contribute/contribute-2015-ga.js:
```diff
@@ -154,9 +164,12 @@ $(function () {
-            event: 'contribute-submit',
-            newsletterSignup: newsletterstate
+            'event': 'contribute-signup-submit',
+            'contributeArea': $('#inquiry-form').data('contribute-area'),
```
Since `$('#inquiry-form')` is being called multiple times, let's cache it in a variable at the beginning of the script like `var $inquiryForm = $('#inquiry-form');`.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -57,16 +57,30 @@ $(function() {
+                        window.dataLayer.push({
+                            'event': 'manifesto-quote-click-arrow',
+                            'direction': ($this.attr('class') === 'prev') ? 'previous' : 'next',
+                            'quote': quoteText.slice(3),
```
Is it a problem that this text will be in different languages? Also, it may not be safe to rely on string positions like this when dealing with localized text.

jpetto replied:
A better approach would be to add `*-data` attributes to each quote defining the quote number and a (preferably shortened) quote. These `*-data` attributes are not localized, meaning we can rely on them always being in English.

JamesDLorence replied:
Are you suggesting hard coding in the values to be added as I don't see a way to dynamically generate this content to add to the data attributes that haven't already been translated?

jpetto replied:
Yeah, I suggest hard coding GA specific values in `*-data` attributes in the markup. (We do this in [some places](https://github.com/mozilla/bedrock/blob/master/bedrock/firefox/templates/firefox/desktop/customize.html#L46) already.)

For example, you could add `data-ga-quote-number` and `data-ga-quote` attributes to each `section.principle` in [manifesto.html](https://github.com/mozilla/bedrock/blob/master/bedrock/mozorg/templates/mozorg/about/manifesto.html#L92). Then, instead of parsing the text of each quote, you could pull the pre-defined `*-data` attributes. (These `*-data` attributes would not be wrapped for L10n, as they would only be used for GA.)

The above strategy may also help keep GA tidy, as you (along with Gareth?) could define short names for each quote instead of using the full quote text, e.g. "enrich-lives" instead of "The Internet must enrich the lives of individual human beings".

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -57,16 +57,30 @@ $(function() {
+
+            window.dataLayer.push({
+                'event': 'manifesto-quote-click',
+                'quote': $(this).find('h3').text().slice(3),
```
Could use `$this` here instead? Also, same warning about localized text - this quote will be in different languages.

jpetto commented on media/js/firefox/family-nav.js:
```diff
@@ -121,6 +122,13 @@ if (typeof window.Mozilla === 'undefined') {
+        $tertiaryNavs.on('click', 'a', function() {
+            window.dataLayer.push({
+                'event': 'family-nav-interaction',
+                'browserAction': $(this).text()
```
This text will be in different languages. Is that okay?

jpetto commented:
A few performance fixes, but nothing too major. Biggest question at this point concerns using localized text in `dataLayer` calls.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -106,10 +120,10 @@ $(function() {
-            browserAction: 'tweet',
-            section: section
+            'event': 'manifesto-quote-share',
+            'quote': $('.principle.overlay-contents h3').text().slice(3)
```
Localized text.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -153,9 +167,9 @@ $(function() {
-                section: section + ': ' + $this.text()
+                'event': 'manifesto-interaction',
+                'browserAction': action,
+                'section': section + ': ' + $this.text()
```
Localized text.

JamesDLorence commented:
@jpetto - I've pushed up the changes. Ready for you to review.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -45,6 +45,8 @@ $(function() {
         $this.attr({
             'tabindex': '0'
         }).on('click', function () {
+            var $this = $(this);
```
Is this necessary? `$this` is already cached above on line 42.

jpetto commented on media/js/mozorg/manifesto.js:
```diff
@@ -57,16 +59,30 @@ $(function() {
+            window.dataLayer.push({
+                'event': 'manifesto-quote-click',
+                'quote': $this.data('ga-quote'),
+                'quoteNumber': $this.attr('data-ga-quote-number')
```
Any reason you're using `.attr` here but `.data` above? Would be better to be consistent if possible.

jpetto commented:
Only a couple small questions/issues. Almost there!

When you next update the branch, please shorten the commit message and reference [the bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1209569) Gareth just created. I'd recommend a commit message like:

`[fix bug 1209569] GTM updates`

Thanks!

JamesDLorence commented:
@jpetto - Changes have been made and pushed and ready for review. Thanks for the patience.

jpetto commented:
Excellent! r+ on the code.

Gareth is finishing the review on demo3. Once he gives the :+1:, we can merge.

jpetto commented:
As this PR is still in progress and will need incremental fixes, please make additional changes in separate commits. This makes it easier to review what's changed since the previous review.

What the code *and* data results are given an r+, we can squash the commits before merge.

Thanks!

JamesDLorence commented:
@jpetto - I think I'm going to need to make an small update. Making a new commit sounds good, does that mean I need to rebase the current commit and then commit on top, or can I just pull on top of the current commit and then commit on all the changes you guys have made?

jpetto commented:
I'd suggest rebasing your branch against mozilla/bedrock/master first to get conflicts resolved. Then make your changes in a new commit(s). The result should be your 2 (or more) commits on top of the current commits on mozilla/bedrock/master.

After that, I'll review your new commit(s), and once they're good, you can squash them into a single commit before we merge.

Sound good?

JamesDLorence commented:
@jpetto - Just pushed a very small addition.

jpetto commented:
@JamesDLorence - Is the latest commit the last of the changes? Or are there more to come?

JamesDLorence commented:
@jpetto - This should be the last. We're doing final rounds of testing now and so I don't expect there to be any more code changes.

JamesDLorence commented:
@jpetto - I added the line `'js/base/mozilla-accordion-gatrack.js',` where you indicated in `static_media.py` but when testing the change I don't see that `mozilla-accordion-gatrack.js` gets added to the page. Is there something extra that I need to do besides spinning up the server to test the change?

jpetto commented:
Hm, that should be it. Can you add another commit so I can see the current code?

JamesDLorence commented:
There you go.

jpetto commented:
Ah, I see what happened. My link on where to add the reference in `static_media.py` was made incorrect due to [a change in that file yesterday](https://github.com/mozilla/bedrock/commit/129f74e9321d13f613c8be79af6670e0f7f8784f#diff-2df771560b784345246f66fa784cce23R544).

You'll need to move the reference to `mozilla-accordian-gatrack.js` from the `firefox_all` JS bundle to the `firefox_android` JS bundle. That should do the trick.

JamesDLorence commented:
That seems to have done it. Thanks! It's ready for review.

jpetto commented:
The changes look good to me! Gareth has confirmed data is coming in as expected, so the last step is to squash these 3 commits into 1.

When you squash, please remove all the old, aggregated commit messages:

![screen shot 2015-10-14 at 2 36 54 pm](https://cloud.githubusercontent.com/assets/317532/10495362/1791c124-7281-11e5-95a7-2ca0017255dd.png)

The only line in the final commit message should be "[fix bug 1209569] GTM updates".

Thanks!

JamesDLorence commented:
Just squashed commits and removed all old commit messages. So glad to be wrapping this up! 

Just to confirm, I haven't done a pull for recent changes in a while. Would you like me to do that and then push to our master?

jpetto commented:
As long as the second check mark is green ("Merging can be performed automatically."), there's no need to rebase. The destination branch hasn't changed so much that the code in this PR can't be merged by a robot. :smile:

Hooray for getting this finished!
